src/UnionSums.hs

```diff
@@ -30,22 +32,44 @@ unionSumTypes newNameStr typeNames =
     return [DataD [] (mkName newNameStr) [] Nothing newConstructors []]
   where
     mkNewConstructors typeName = do
-      info <- reify typeName
-      case info of
-        (TyConI (DataD _ _ _ _ constructors _)) ->
-          either fail return $ mapM (modConstructor typeName) constructors
-        _ -> fail $ nameBase typeName ++ " must be a sum type"
+      cons <- getConstructors typeName
+      either fail return $ mapM (modConstructor typeName) cons
 
     modConstructor :: Name -> Con -> Either String Con
     modConstructor typeName (NormalC conName args) =
-        (\n -> (NormalC (mkName $ n ++ newNameStr) args)) <$>
+        (\n -> NormalC (mkName n) args) <$>
         note "Constructor name missing type name suffix"
-          (stripNameSuffix typeName conName)
+          (changeSuffix (nameBase typeName) newNameStr (nameBase conName))
     modConstructor _ _ = fail "Unrecognised constructor pattern"
 
-    stripNameSuffix :: Name -> Name -> Maybe String
-    stripNameSuffix suf = stripSuffix (nameBase suf) . nameBase
+mkConverter :: Name -> Name -> Q [Dec]
+mkConverter subName mainName = do
+    ft <- [t| $(conT subName) -> $(conT mainName) |]
+    cs <- getConstructors subName >>= mapM toClause
+    return [SigD name ft, FunD name cs]
+  where
+    name = mkName $ lowerFirst $ (nameBase subName) ++ "To" ++ (nameBase mainName)
+    swapSuffix n = mkName $ fromJust $ changeSuffix (nameBase subName) (nameBase mainName) $ nameBase n
+    toClause (NormalC conName args) = do
+        argNames <- mapM (const $ newName "a") args
+        return $ Clause (conPattern conName argNames) (recon (swapSuffix conName) argNames) []
+    conPattern conName argNames = [ConP conName $ map VarP argNames]
+    recon mainConName argNames = NormalB $ foldl AppE (ConE mainConName) $ map VarE argNames
+
+getConstructors :: Name -> Q [Con]
+getConstructors n = do
+    info <- reify n
+    case info of
+        (TyConI dec) -> case dec of
+            (DataD _ _ _ _ cs _) -> return cs
+            _ -> fail $ (nameBase n) ++ " does not only contain data definitions"
+        _ -> fail $ (nameBase n) ++ " must be a sum type"
 
+changeSuffix :: String -> String -> String -> Maybe String
```

Comment:
Every time you've used this you've converted the `String`s to `Name`s. So, you could factor that out:
```haskell
changeNameSuffix :: Name -> Name -> Name -> Maybe Name
changeNameSuffix oldSuffix newSuffix = mkName . changeSuffix (nameBase oldSuffix) (nameBase newSuffix) . nameBase
```
Then your usage site in the where clause becomes:
```haskell
swapSuffix n = fromJust $ changeNameSuffix subName mainName n
```
or just
```haskell
swapSuffix = fromJust . changeNameSuffix subName nameBase
```